### PR TITLE
Add tt crud-import module

### LINKED ISSUE #35: Implement `tt crud import`
Implement the `import` functionality in the `crud` module to import data from CSV files into tarantool via [`crud`](https://github.com/tarantool/crud).

Seems like the following components must be implemented as part of this task:
* common interface of the import module
* CSV parser
* `crud` implementation of the import module

The CLI options were described in https://github.com/tarantool/tt/issues/24 .

### PR DESCRIPTION
**Pull request by GRISHNOV**
Add a module responsible for importing the contents from file to Tarantool via crud.

Closes #35

### CODE REVIEW

File: cli/ttparsers/ttcsv.go [hunk 1]
```diff
@@ -0,0 +1,482 @@
+// Copyright 2011 The Go Authors. All rights reserved.
```
Comment (LeonidVas) [inline cli/ttparsers/ttcsv.go:1]: Move the csv parser addition to a separate commit.

File: cli/cmd/crudimport.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/apex/log"
+	"github.com/spf13/cobra"
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/connector"
+	crudimport "github.com/tarantool/tt/cli/crud/import"
+	"github.com/tarantool/tt/cli/modules"
+	ttcsv "github.com/tarantool/tt/cli/ttparsers"
+)
+
+// crudImportFlags contains flags for cat command.
```
Comment (LeonidVas) [inline cli/cmd/crudimport.go:16]: for **cat** command - No)

File: cli/crud/import/import.go [hunk 3]
```diff
@@ -0,0 +1,1214 @@
+/*
+	The batches are transmitted to the router in the form of a serialized json (struct t_batch).
+	Feedback from crud is obtained through this way also.
+
+	At the end of the program batches amount will be ⌈csvRecodrdsAmount / batchSize⌉, where csvRecodrdsAmount
```
Comment (LeonidVas) [inline cli/crud/import/import.go:5]: Here and below the maximum line length is 100.

File: cli/crud/import/import.go [hunk 4]
```diff
@@ -0,0 +1,1214 @@
...
+	"time"
+
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/connector"
+	ttcsv "github.com/tarantool/tt/cli/ttparsers"
+	"github.com/tarantool/tt/cli/util"
+	"gopkg.in/yaml.v2"
+)
+
+// t_importSummary contains information for the import summary.
```
Comment (LeonidVas) [inline cli/crud/import/import.go:34]: What does `t_` mean? In golang we (and not only) do not use snake_case.
Here and below.

File: cli/crud/import/import.go [hunk 5]
```diff
@@ -0,0 +1,1214 @@
...
+	readTotal       uint32
+	parsedSuccess   uint32
+	parsedError     uint32
+	importedSuccess uint32
+	importedError   uint32
+}
+
+// Opts contains information about options from the user via the CLI.
+type Opts struct {
+	ConnectUsername string
```
Comment (LeonidVas) [inline cli/crud/import/import.go:45]: Add description for all public (at least) attributes.

File: cli/crud/import/import.go [hunk 6]
```diff
@@ -0,0 +1,1214 @@
...
+
+	"github.com/tarantool/tt/cli/cmdcontext"
+	"github.com/tarantool/tt/cli/connector"
+	ttcsv "github.com/tarantool/tt/cli/ttparsers"
+	"github.com/tarantool/tt/cli/util"
+	"gopkg.in/yaml.v2"
+)
+
+// t_importSummary contains information for the import summary.
+type t_importSummary struct {
```
Comment (LeonidVas) [inline cli/crud/import/import.go:35]: Maybe it would be nice to move all `types` to a separate file to make code reading more convenient?

File: cli/crud/import/import.go [hunk 7]
```diff
@@ -0,0 +1,1214 @@
...
+	currentRec := ctx._currentRecord
+	ctx._currentRecord = ""
+
+	return currentPosition, currentRec
+}
+
+var importSummary t_importSummary = t_importSummary{}
+
+// RunImport implements data import via crud.
+func RunImport(cmdCtx *cmdcontext.CmdCtx, conn *connector.Conn, csvReader *ttcsv.Reader, rawReaderFd *os.File, crudImportFlags *Opts, spaceName string) error {
```
Comment (LeonidVas) [inline cli/crud/import/import.go:183]: The function ~ 600 lines it's not cool. MOve some part of logic to  a separated functions.

File: cli/crud/import/import.go [hunk 8]
```diff
@@ -0,0 +1,1214 @@
...
+	sigIntInput := make(chan os.Signal, 1)
+	signal.Notify(sigIntInput, os.Interrupt, syscall.SIGTERM)
+	sigIntInterceptor(sigIntInput)
+
+	// init progress file context
```
Comment (LeonidVas) [inline cli/crud/import/import.go:188]: https://github.com/tarantool/tarantool/wiki/Code-review-procedure#code-style
> Start sentences from a capital letter, end with a dot.

Here and below.

File: cli/crud/import/import.go [hunk 9]
```diff
@@ -0,0 +1,1214 @@
...
+
+	fmt.Printf("In case of error:\t[%s]\n", crudImportFlags.OnError)
+	fmt.Printf("PID of this process:\t[%d]\n", os.Getpid())
+	fmt.Printf("\n\033[0;33mWARNING: Process is not sensitive to SIGINT (ctrl+c), use kill -9 %d\033[0m\n", os.Getpid())
+
+	var batchSequenceCtx *t_batchSequenceCtx
+	var batch *t_batch
+	if crudImportFlags.Header {
+		/*
+			initialization in the case of a header use
```
Comment (LeonidVas) [inline cli/crud/import/import.go:242]: Use `//' for comments everywhere.
Here and below.

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Hi! Thank yo for the patch.
See the several common comments.
I will following to review the code.

File: cli/crud/import/import.go (hunk 9 above)
Comment (GRISHNOV) [reply cli/crud/import/import.go]: Done

File: cli/crud/import/import.go (hunk 8 above)
Comment (GRISHNOV) [reply cli/crud/import/import.go]: Done

File: cli/crud/import/import.go (hunk 6 above)
Comment (GRISHNOV) [reply cli/crud/import/import.go]: Moved to `cli/crud/import/containers.go`

File: cli/crud/import/import.go (hunk 5 above)
Comment (GRISHNOV) [reply cli/crud/import/import.go]: Done

File: cli/crud/import/import.go (hunk 4 above)
Comment (GRISHNOV) [reply cli/crud/import/import.go]: Postfix as an abbreviation of the word `type`
Now using postfix `Container` for naming structs; defined all related types in `cli/crud/import/containers.go`

File: cli/crud/import/import.go (hunk 3 above)
Comment (GRISHNOV) [reply cli/crud/import/import.go]: Done

File: cli/cmd/crudimport.go (hunk 2 above)
Comment (GRISHNOV) [reply cli/cmd/crudimport.go]: Done

File: cli/ttparsers/ttcsv.go (hunk 1 above)
Comment (GRISHNOV) [reply cli/ttparsers/ttcsv.go:1]: Done

File: README.rst [hunk 10]
```diff
@@ -176,3 +176,4 @@ Common description. For a detailed description, use ``tt help command`` .
 * ``rocks`` - LuaRocks package manager.
 * ``cat`` - print into stdout the contents of .snap/.xlog files.
 * ``play`` - play the contents of .snap/.xlog files to another Tarantool instance.
+* ``crud-import`` - import data from file into Tarantool via crud.
```
[the remark this answers is not in the document]
Comment (GRISHNOV) [reply README.rst]: Done

File: CHANGELOG.md [hunk 11]
```diff
@@ -23,3 +23,4 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Module ``tt rocks``, LuaRocks package manager.
 - Module ``tt cat``, to print into stdout the contents of .snap/.xlog files.
 - Module ``tt play``, to play the contents of .snap/.xlog files to another Tarantool instance.
+- Module ``tt crud-import``, to import data from file into Tarantool via crud.
```
[the remark this answers is not in the document]
Comment (GRISHNOV) [reply CHANGELOG.md]: Done

**Comment by GRISHNOV** [comment]
Thank you for your feedback! I tried to answer all the comments on the code review

File: cli/crud/import/import.go (hunk 7 above)
Comment (GRISHNOV) [reply cli/crud/import/import.go]: At the moment, the largest function `mainParsingCycle` is ~440 rows. I don't find an opportunity to improve the result because of the operators `continue` and `break` in this func

File: cli/cmd/crud.go [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+	if crudImportFlags.BatchSize == 0 {
+		return fmt.Errorf("The batch size must be greater than zero.")
+	}
+
+	if crudImportFlags.OnError != "skip" && crudImportFlags.OnError != "stop" {
+		return fmt.Errorf("The option onerror can be <skip> or <stop>.")
+	}
+
+	if (crudImportFlags.Match == "header" && !crudImportFlags.Header) ||
```
Comment (LeonidVas) [inline cli/cmd/crud.go:126]: If `header` is true, what for to specify it in `match`?

File: cli/cmd/crud.go [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+		(crudImportFlags.Match != "" && crudImportFlags.Match != "header") {
+		return fmt.Errorf("Currently only <header> value supported for match option. " +
+			"Also it require option header as <true>.")
+	}
+
+	csvReaderFd, err := os.Open(args[1])
```
Comment (LeonidVas) [inline cli/cmd/crud.go:132]: Add a comment, why we open the one file twice. (`csvReaderFd`, `rawReaderFd`).

File: cli/crud/import/containers.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
+// This file contains types for defining containers (structs) that store
+// contexts during crud import work.
+
+package crudimport
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:4]: Why do we need two packages (`crudimport` and `crudexport`)? I propose to just have a package `crud`.

File: cli/crud/import/containers.go [hunk 15]
```diff
@@ -0,0 +1,163 @@
+// This file contains types for defining containers (structs) that store
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:1]: Why `containers', not `types'?

File: cli/crud/import/containers.go [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+package crudimport
+
+import (
+	"bufio"
+	"os"
+)
+
+// importSummaryСontainer contains information for the import summary.
+type importSummaryСontainer struct {
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:12]: `importSummaryСontainer` -> `importSummary`

File: cli/crud/import/containers.go [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+type crudCtxСontainer struct {
+	// TupleImported indicates the success of the import.
+	TupleImported bool `yaml:"tupleImported" json:"tupleImported"`
+	// CastedTuple сontains a set of converted (on router side) fields to insert into space.
+	CastedTuple []interface{} `yaml:"castedTuple"  json:"castedTuple"`
+	// CrudErr сontains error description from crud stored procedure.
+	CrudErr string `yaml:"crudErr" json:"crudErr"`
+}
+
+// batchTupleСontainer contains the context of a tuple within a batch.
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:93]: `batchTupleСontainer` -> `batchCtx`

File: cli/crud/import/containers.go [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+type batchTupleСontainer struct {
+	// TupleNumber contains the number of the tuple within the batch.
+	// It can take values from the range [1, batchSize] for initialized tuples within a batch.
+	// For uninitialized tuples within a batch, a 0 value is used.
+	// The current number of initialized tuples is recorded in struct batchСontainer->TuplesAmount.
+	TupleNumber uint32 `yaml:"tupleNumber" json:"tupleNumber"`
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:99]: `TupleNumber` -> `Number`?

File: cli/crud/import/containers.go [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+	logFile *os.File
+	// errorFile is file descriptor of file with unsuccessful imported data.
+	errorFile *os.File
+	// successFile is file descriptor of file with successful imported data.
+	successFile *os.File
+	// progressFile is file descriptor of file with progress information.
+	progressFile *os.File
+}
+
+// parserCtxСontainer contains a part of the tuple context that related to the parser.
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:69]: I think it doesn't matter what the structure is a part of, what matters is what it describes.

File: cli/crud/import/containers.go [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+	// For uninitialized tuples within a batch, a 0 value is used.
+	// The current number of initialized tuples is recorded in struct batchСontainer->TuplesAmount.
+	TupleNumber uint32 `yaml:"tupleNumber" json:"tupleNumber"`
+	// ParserCtx contains parser context related to this tuple.
+	ParserCtx parserCtxСontainer `yaml:"parserCtx" json:"parserCtx"`
+	// CrudCtx contains router side context related to this tuple.
+	CrudCtx crudCtxСontainer `yaml:"crudCtx" json:"crudCtx"`
+}
+
+// batchСontainer contains the context of the batch, including necessary meta information for crud.
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:106]: The `BatchContainer' and `BatchTruleContainer' are very close names. I suggest you think about the names.

File: cli/crud/import/containers.go [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+	// Header сontains a header of input data.
+	Header []string `yaml:"header" json:"header"`
+	// TuplesAmount сontains the number of initialized tuples within batch.
+	TuplesAmount uint32 `yaml:"tuplesAmount" json:"tuplesAmount"`
+	// Tuple is slice of tuples within this batch.
+	Tuple []batchTupleСontainer `yaml:"tuples" json:"tuples"`
+}
+
+// progressCtxСontainer contains the context for the progress file.
+type progressCtxСontainer struct {
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:119]: `progressCtxСontainer` -> `progressCtx`

File: cli/crud/import/containers.go [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+	// StartTimestamp contains the work start time of this launch.
+	StartTimestamp string `yaml:"startTimestamp" json:"startTimestamp"`
+	// LastDumpTimestamp contains the time of last write to progress file of this launch.
+	LastDumpTimestamp string `yaml:"lastDumpTimestamp" json:"lastDumpTimestamp"`
+	// EndOfFileReached indicates has the END been reached or not.
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:124]: `EOF`?

File: cli/crud/import/containers.go [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+	LastPosition uint32 `yaml:"lastPosition" json:"lastPosition"`
+	// RetryPosition contains positions of input file that could not be imported in this launch.
+	RetryPosition []uint32 `yaml:"retryPositions" json:"retryPositions"`
+	// lastPositionPrevProgress contains LastPosition from previous version of progress file.
+	lastPositionPrevProgress uint32
+	// retryPositionPrevProgress contains RetryPosition from previous version of progress file.
+	retryPositionPrevProgress []uint32
+}
+
+// batchSequenceCtxСontainer contains the context for
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:136]: `batchSequenceCtxСontainer` -> `batchSequenceCtx`
Also we have `batchСontainer` and `batchTupleСontainer`. Ouch...

File: cli/crud/import/containers.go [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+// unparsedCsvReaderCtxСontainer contains private fields for organizing context of
+// raw reading of input data for import.
+// Raw reading is necessary to get a string of data in an unparsed form.
+// The standard parser does not allow this.
+// It also contains fields for implementing a mechanism to eliminate
+// the inconsistency of the position counter due to a multi-line CSV records.
+// Fields inside this structure should not be read or overwritten by the programmer.
+// Exclusion are methods updCtxAfterParseError, updCtxAfterRecordHandling.
+type unparsedCsvReaderCtxСontainer struct {
+	_masterPosition uint32
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:159]: In golang, fields starting with lowercase letters are private, do not use underscores here.

File: cli/crud/import/containers.go [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+
+// unparsedCsvReaderCtxСontainer contains private fields for organizing context of
+// raw reading of input data for import.
+// Raw reading is necessary to get a string of data in an unparsed form.
+// The standard parser does not allow this.
+// It also contains fields for implementing a mechanism to eliminate
+// the inconsistency of the position counter due to a multi-line CSV records.
+// Fields inside this structure should not be read or overwritten by the programmer.
+// Exclusion are methods updCtxAfterParseError, updCtxAfterRecordHandling.
+type unparsedCsvReaderCtxСontainer struct {
```
Comment (LeonidVas) [inline cli/crud/import/containers.go:158]: `unparsedCsvReaderCtxСontainer` -> `unparsedCsvReaderCtx`.
Maybe just `csvReaderCtx`?

**Review by LeonidVas — CHANGES_REQUESTED** [review]

File: cli/crud/lua/import.lua [hunk 26]
```diff
@@ -0,0 +1,465 @@
...
+local init_eval_set_targetspace = function()
+    box.session.storage.crudimport_set_targetspace = function(space)
+        box.session.storage.crudimport_targetspace = space
+    end
+end
+
+-- Check existance of target space for import.
+local init_eval_check_targetspace_exist = function()
+    box.session.storage.crudimport_check_targetspace_exist = function()
+        if crud.select(box.session.storage.crudimport_targetspace) ~= nil then
```
Comment (Totktonada) [inline cli/crud/lua/import.lua:76]: Here you'll fetch the whole space, but AFAIU you need only metadata. Please, limit the select with `first = 0`.

File: cli/crud/lua/import.lua [hunk 27]
```diff
@@ -0,0 +1,465 @@
...
+            box.session.storage.crudimport_targetspace_format =
+                crud.select(box.session.storage.crudimport_targetspace)['metadata']
```
Comment (Totktonada) [inline cli/crud/lua/import.lua:78]: You already makes this request a line above. It is better to save the result and reuse it instead of sending the request twice.

File: cli/crud/lua/import.lua (hunk 26 above)
Comment (GRISHNOV) [reply cli/crud/lua/import.lua]: Done

File: cli/crud/lua/import.lua (hunk 27 above)
Comment (GRISHNOV) [reply cli/crud/lua/import.lua]: Done

File: cli/crud/lua/import.lua [hunk 28]
```diff
@@ -0,0 +1,467 @@
...
+                    then
+                        batch_final_ctx['tuples'][parsed_tuple_num]['crudCtx']['imported'] = true
+                        -- Try to resist ambiguity, but a swap is still possible (at least we don't lose record).
+                        box.session.storage.batch_insert_res['rows'][imported_tuple_num] = nil
+                        goto resulted_tuple_found
+                end
+            end
+            for error_tuple_num, error_tuple_val in pairs(box.session.storage.batch_insert_err) do
+                if box.session.storage.crud_import_is_tuples_equal(
+                    error_tuple_val['tuple'],
```
Comment (Totktonada) [inline cli/crud/lua/import.lua:430]: It was renamed to `operation_data` in the recent `crud.*_many()` implementation, because it may be an object for `crud.*_object_many()`. Please, update accordingly.

Comment (GRISHNOV) [reply cli/crud/lua/import.lua:430]: Done
